### PR TITLE
Standardized API for external communication

### PR DESCRIPTION
**Pull request by BernhardPosselt**
WIP for #15894 

:construction:

### CODE REVIEW

**Comment by Raydiation** [comment]
@DeepDiver @pucela @butonic @MorrisJobke @rullzer

**Comment by DeepDiver1975** [comment]
Thanks a lot for pushing this foreward!

I order to really settle this to be used everywhere we have to face one major migration issue: 
how to integrate/migrate the existing public OCS API?

It is key to keep the existing API routes in addition to have no code duplication.

We might want to use http redirect from the individual old ocs routes to the new api routes?
Using rewrite rules feels a bit too unreliable.

Furthermore what will be the route to get the json file with all existing APIs?
OCS has a similar construct - we shall merge/combine the effort.

**Comment by Raydiation** [comment]
@DeepDiver1975 let us handle OCS stuff in a different PR. Once all features in #15894 are in, we can merge this PR and build OCS compatibility on top of it. 

Current focus of this PR is to create an api routing infrastructure that does all the security stuff.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 let us handle OCS stuff in a different PR. Once all features in #15894 are in, we can merge this PR and build OCS compatibility on top of it.

second pr: for sure
thinking about OCS compatibility afterwards: no really - we need the solution at hand before we go further - otherwise we will just create another api and it will get even more messy

**Comment by Raydiation** [comment]
Redirects is the obvious solution here then

**Comment by DeepDiver1975** [comment]
> Redirects is the obvious solution here then

@PVince81 @LukasReschke objections?

**Comment by Raydiation** [comment]
Will add the same for "old" news apis btw.

**Comment by PVince81** [comment]
No objections.

On a side note I'm still kind of stuck in the Webdav vs OCS discussion... but moving forward is good too.

**Comment by Raydiation** [comment]
btw @PVince81 using webdav for the webinterface is only possible if your controllers serve webdav responses, otherwise you've got CSRF issues.

**Comment by MorrisJobke** [comment]
@Raydiation : @pucela is now @cmonteroluque ;)

**Comment by Raydiation** [comment]
@DeepDiver1975 got a second idea about the ocs api migration stuff. What if we implemented some kind of internal route rewrite logic?

The News app has the following problem: The current api URL is 
    
    https://yourowncloud.com/index.php/apps/news/api/v1-2/

If the API is changed to the new implementation it will most likely become

    https://yourowncloud.com/index.php/api/news 

Putting the version into a separate header seems to be a more robust idea :)


So the problem with this change is the following:
* Either I break backwards compatilibity (which would be horrible) and wait for everyone to migrate to the new base url. This is not an option
* I create nearly empty controllers which all return redirect responses. The question is how well clients will support this and if this works across http verbs.

Since all we need is wrapping old URLs to new URLs to migrate and still provide the old API, we could simply define internal route redirects. An example would be the following:

<?php
class Routes {
    const WEB = 0;
    const API = 1;
    const OCS = 2;
}

return [
    // standard stuff
    'resources' => [..],
    'routes' => [..],
    'api_resources' => [..],
    'api_routes' => [..],
    
    'rewrites' => [
        [
            'from' => ['url' => '/api/v1-2/{path}', 'endpoint' => Routes::WEB, /* optionally 'app' => 'news' */],
            'to' => ['url' => '/$1', 'endpoint' => Routes::API, /* optionally 'app' => 'news' */],
            'requirements' => ['path' => '.*']
        ]
    ]
];

**Comment by Raydiation** [comment]
Committed some work in progress security profiles. Depending on if the request was made from an API route or @CORS annotated controller method, an API profile will be chosen, otherwise the web profile will run the required checks.

An additional goal is to merge the security middleware with the cors middleware and to only handle the http stuff in there, security checks are done in different classes.

I'm currently wondering what https://github.com/owncloud/core/blob/master/lib/private/appframework/middleware/security/securitymiddleware.php#L132 is for @LukasReschke 

Why do we have to check if an app is enabled?

**Comment by Raydiation** [comment]
If an app is not enabled, shouldnt the routes.php and app.php file not be loaded anyways? Meaning an app which is not enabled won't ever execute code?

**Comment by Raydiation** [comment]
Stuff which still has to be done
* [ ] merge cors and security middlware
* [ ] adjust tests for security middlware and profiles

File: lib/private/appframework/http/request.php [hunk 1]
```diff
@@ -400,29 +400,33 @@ protected function getContent() {
...
+	 * string
+	 */
+	public function getRequestCSRFToken() {
+		$possibleTokenLocations = [
+			'get' => 'requesttoken',
+			'post' => 'requesttoken',
+			'server' => 'HTTP_REQUESTTOKEN'
+		];
+
+		$token = '';
```
Comment (Raydiation) [inline lib/private/appframework/http/request.php]: @LukasReschke can you check if its ok to simply return an empty string for the request token if not present? I'd like to keep the return type and not juggle around with null or bool values :)

File: lib/private/appframework/http/request.php [hunk 2]
```diff
@@ -698,4 +702,12 @@ private function getOverwriteHost() {
 		return null;
 	}
 
+	/**
+	 * Returns true if the request hits an API route, e.g. index.php/api/myapp
+	 * @return bool
+	 */
+	public function isApiRequest() {
+		return strpos($this->getPathInfo(), '/api') === 0;
```
Comment (Raydiation) [inline lib/private/appframework/http/request.php]: @LukasReschke is this bullet proof?

**Comment by PVince81** [comment]
Defer to 9.0 ?

**Comment by DeepDiver1975** [comment]
> Defer to 9.0 ?

yes

**Comment by DeepDiver1975** [comment]
Sorry - I don't see this leading us anywhere with the current state.

**Comment by DeepDiver1975** [comment]
I'm closing this for now

**Pull request closed without merging**